README.md
# docker-TimImage_builder rel1.0-23

**Purpose**: Creates configuration files to create CA APM TIM docker Images and Containers  
  _by J. Mertin -- joerg.mertin(-AT-)ca.com_

## Description
The docker-TimImage_builder permits to install a TIM from version 9.6.x to
10.x practically on any Docker Server in a matter of minutes.

## Short description
The docker-TimImage_builder permits to install a CA APM TIM from version
9.6.x to 10.x practically on any Docker Server in a matter of minutes.
The build-script provides a first-configuration UI, and will also
create a script to start/stop the TIM docker container later with all
pre-configured data inside.

## APM version
So far - the docker-builder has been tested with TIM from version 9.6
up to 10.5.

## Installation Instructions

#### Prerequisites
- Docker-Server for hosting the container
- Dedicated network interface for the SPAN Provider. This Network
  interface will be dedicated to the TIM Container - and cannot be
  accessed by any other applicaiton, container after the TIM container
  is activated!
- Copy of the TIM installation files for RedHat/CenOS. Download from
  CA Support site
- The ca-eula.en.txt file needs to be edited and content accepted!

#### Dependencies
- None - the container setup will provide all necessary dependencies


#### Configuration
The configuration section has been set forward, as it is done prior
the docker file creation, hence before the container is created and/or
started.  The `TIMImage_builder.sh` script, when started without
configuration file as argument will interactively ask the information
required to run the TIM inside the docker environment. The regular TIM
configuration will not be impacted by this.


#### Installation

First of all you need to get a copy of the repository or simply clone
it. Get the link - and clone the repository:

`~# git clone https://github.com/CA-APM/docker-TimImage_builder.git`

The TIM Version the container will run is defined by the files
provided in the timXX.x directory. The maker of the Docker-Image needs
to download the TIM files from CA Support site (Download section),
untar the version for CentOS/RedHat 6.x and place the obtained files
into the appropriate version directory, example "tim10.1":

```
admin@antigone:~/docker/docker-TimImage_builder$ ls -l tim10.2/*  
-rwx------ 1 admin adm     1698 Mar 17 01:43 tim10.2/CA-APM-TIM-public_key.txt  
-rw-r--r-- 1 admin adm   352952 Jun 15 10:40 tim10.2/ca-eula.en.txt  
-rwxr-xr-x 1 admin adm 11470907 Mar 16 12:58 tim10.2/timInstall.bin
```

You can have more than one version in the base directory. Make sure to
edit the ca-eula.en.txt file and set the Eula Acceptance to
**CA-EULA=reject** or the installation of the TIM will fail.

This docker-TimImage_builder provides a CLI UI to choose the initial
setup of the TIM docker build, letting the user choose between
different versions, set the name of the image-build and the container
name. In case of the secured network mode, will set the ports that
will be exposed. By default - the TIM docker will use the internal
ports 80, 81, 8080 and 8433, which eventually need to be remapped.  In
"host" network mode, the current network configuration of the host
will be made available to the docker container. This will be very
practical in case one wants to monitor a webserver running on the same
hardware.  
_Note: in case you want to access the container using https, you will
have to provide a valid private key and certificate. Else the server
will refuse secured accesses._


#### For Interactive mode, just call:  
`~# ./TIMImage_builder.sh`

_Note: Every interactive mode will create a configuration file in the
"cfgs" directory. You can then use it for the *automated* mode._

#### Automated mode: provide a configuration file as argument  
`~# ./TIMImage_builder.sh cfgs/tim9.7.0.cfg`

In both cases, the script will show the current settings and ask for
confirmation of the image to be built.

```
>>> Options for this Docker Image Build
===============================================================
TIM Version:            tim10.2
TIM Workers:            2
Image name:             caapm/tim10.2
Exposed ports:          8080 8443 81 80
Network mode:           secured
Container Name:         tim10.2
Host SPAN Int.:         dummy0
Port 80 mapped to:      80
Port 81 mapped to:      81
Port 8080 mapped to:    8080
Port 8443 mapped to:    8443

===============================================================
>>> Proceed [y/n]:
```

The Docker-Builder will then create some files that are used to create
the actual docker image, and provide a start script to start the
Container, and a script to enter the container.

The TIM software will be installed in unattended mode using the defaults
as provided by the **CA_AUTOMATION**. 


#### Usage Instructions

Every time the `TIMImage_builder.sh` is executed, a CentOS 6.8 OS
image is built, a complete RPM update is performed before the actual
TIM installation, providing an up to date and secure environment
for the TIM process to run in.

The main issue for the TIM to actually run inside a Docker
environment, is to provide apmpacket the data-flow the TIM workers are
supposed to monitor. Docker however has been designed to share
resources. This is the reason the Admin starting the the TIM Container
will have insert the SPAN Device into the Docker-Container, or use the
generated `start_tim.sh` shell script so the container can directly
"hijack" the defined SPAN interfaces on the running host.

The `TIMImage_builder.sh` execution will create 2 scripts:

- **start_tim9.7.0.sh**: This script will start the docker container as
  per the name.  

  _Note: in *secured* mode, the configured SPAN interface will be
   hijacked from the docker host and made exclusively available to the
   inside of the TIM container. The host mode will share the host
   network inside the container._
   * In case the container does not yet exist, it will be built and
     started.
   * In case it already exists, the existing instance will be started
     and networking setup according to request.

- **tim9.7.0_shell.sh**: This one will drop you to a shell directly inside
  the running container
  
The TIM Container will create various directories under the user's home
 invoking the start of the container:

| Inside the container | In the user's home directory |
|----------------------|------------------------------|
| /opt/CA/APM/tim/config | ~/docker/{CONTAINER}/config |
| /opt/CA/APM/tim/logs | ~/docker/{CONTAINER}/logs |
| /root/host_dir | ~/docker/{CONTAINER}/exchange |

These directories are persistent. This means that in case you create a
new TIM container with the same release, this directory's content will
not be overwritten.

If you want to upgrade the TIM while keeping the current
configuration, all you need to do is create the new TIM image and
adapt the "{CONTAINER}" to reflect the name of the next running
instance. Do the upgrade in this specific order - example based on
tim10.1 to tim10.3:

1. Create the new Image version tim10.3
2. stop the current image tim10.1 with `docker stop tim10.1`
3. mv the old configuration directory to the new one with: `mv tim10.1 tim10.3`
4. start the new container with the provided start-script: `./start_tim10.3.sh`


## Limitations
* Currently, the container-start script using the secured network mode
  will only work on a Linux Docker Server - as the syntax used to
  Hijack the SPAN Interface is specific to each OS.

* In case the "secured" network mode is used, the network interface
  used for SPAN Port is exclusively assigned to a docker container and
  cannot be used by any other process/docker container.

## Debugging and Troubleshooting
Troubleshooting needs to be done as on a regular TIM.  
_Note: for entering the TIM Container, use the provided `tim9.7.0_shell.sh` script._


## Security

The TIM Container provides no active security. However, during
installation, the latest fixes for the underlying OS are applied. At
the same time - a cron-job is installed to update the underlying OS on
a regular base.

In case this behavior is not wanted, or the container has no access to
regular centos repositories, please remove the `yum.cron` file, and
restart crond.  
```
[root@antigone cron.d]# rm -f /etc/cron.d/yum.cron 
[root@antigone cron.d]# service crond reload
```  
_Note: removing crond alltogether can be done but is not advised as
it also takes care of all the daily/weekly/monthly system logfile
maintenance._


## Legal notice: binaries not included
As CA APM TIM is commercial software I cannot provide fully-built
images. Also this project cannot provide the binaries for building the
images. But the addition of the binaries is easy.


## License
This field pack is provided under the [Eclipse Public License, Version
1.0](https://github.com/CA-APM/docker-TimImage_builder/blob/master/LICENSE).

## Support
This document and associated tools are made available from CA
Technologies as examples and provided at no charge as a courtesy to
the CA APM Community at large. This resource may require modification
for use in your environment. However, please note that this resource
is not supported by CA Technologies, and inclusion in this site should
not be construed to be an endorsement or recommendation by CA
Technologies. These utilities are not covered by the CA Technologies
software license agreement and there is no explicit or implied
warranty from CA Technologies. They can be used and distributed freely
amongst the CA APM Community, but not sold. As such, they are
unsupported software, provided as is without warranty of any kind,
express or implied, including but not limited to warranties of
merchantability and fitness for a particular purpose. CA Technologies
does not warrant that this resource will meet your requirements or
that the operation of the resource will be uninterrupted or error free
or that any defects will be corrected. The use of this resource
implies that you understand and agree to the terms listed herein.

Although these utilities are unsupported, please let us know if you
have any problems or questions by adding a comment to the CA APM
Community Site area where the resource is located, so that the
Author(s) may attempt to address the issue or question.

Unless explicitly stated otherwise this field pack is only supported
on the same platforms as the regular APM CEM TIM. See [APM
Compatibility Guide](http://www.ca.com/us/support/ca-support-online/product-content/status/compatibility-matrix/application-performance-management-compatibility-guide.aspx).


### Support URL
https://github.com/CA-APM/docker-TimImage_builder/issues


## Categories
Packaged Applications Monitoring Virtualization/Containers




CHANGELOG
=== CHANGELOG ==================================================
Fri Jun 24 14:53:21 CEST 2016
	- Added support for host network mode.

Wed Jun 22 16:33:57 CEST 2016
	- Added support for persistent config and logs directories.

Thu Jun 16 09:25:23 CEST 2016
	- Adapted the README.md to CA marketplace requirements

Wed Jun 15 10:35:23 CEST 2016
	- Made the start-script aware of missing SPAN interface and
	  already existing docker container of the same name
	- Added License and Support information
	- Moved from README to README.md
	- Added new container detection and interface handling.

Mon Mar  7 15:27:10 CET 2016
	- removed TIM Binaries and adapted the code to handle it.

Fri Jan 29 10:06:17 CET 201
	- Supported TIM Versions 9.6 to 10.1
	- Initial releasebuild

=== GIT CHANGELOG ==============================================
* 2016-07-20 - (HEAD -> master, tag: v1.0-22, origin/master) Created 2 Dockerfile templates secure and host, to avoid configuring the httpd Listen port 
* 2016-07-20 - Fixed a race condition, where HTTP export ports where also adapted in secure mode (not required, as all is handled through the proxies) 
* 2016-06-28 - Spellchecking fixes 
* 2016-06-28 - Replaced < Tags 
* 2016-06-28 - Updated documentation, added Security section 
* 2016-06-28 - Added port-selection for TIM in host-mode. 
* 2016-06-28 - Adapted the httpd configuration inside of the container 
* 2016-06-24 - (tag: v1.0-21) Added network/host mode option, updated documentation to reflect the changes. 
* 2016-06-24 - Added support for Docker host mode. 
* 2016-06-24 - Added support for host network mode 
* 2016-06-22 - HTML tags don't work in README.md file. 
* 2016-06-22 - Updated documentation 
* 2016-06-22 - (tag: v1.0-20) Added consistent config and logs directory 
* 2016-06-22 - Added pre-tim to be added for persistent config and log file directory to be accessible by admin from outside the container 
* 2016-06-22 - Added persistent directories for config and log files. 
* 2016-06-20 - Updated docs that tim 10.3 works also nicely. 
* 2016-06-17 - Updated docker link 
* 2016-06-17 - (tag: v1.0-19) Initial docker-TimImage_builder git commit 

mod/clean_rc.sh
#!/bin/sh

if [ -f /.CLEANRC ]
then
    exit 0
fi

# Extract Run level
Rlvl=`grep "^id:" /etc/inittab | cut -d ':' -f 2`

for link in `find /etc/rc${Rlvl}.d/ -type l`
do
    sfile="`basename $link`"
    if [ `echo $sfile | grep -c "^S"` -gt 0 ]
    then
	service=`echo $sfile | sed -e 's/^S[0-9][0-9]//g'`
	case $service in
	    "httpd") 
		echo "Leave $service"
		;;
	    "tim") 
		echo "Leave $service"
		;;
	    "tim-assist") 
		echo "Leave $service"
		;;
	    "crond") 
		echo "Leave $service"
		;;
	    "local") 
		echo "Leave $service"
		;;
	    "apmpacket") 
		echo "Leave $service"
		;;
	    *) echo "Disable $service"
		chkconfig $service off
		;;
	esac
    fi
done

# Clean up some more things. We need to disable whatever service that changes internals on this running CentOS
# Prevent Getty's from starting
sed -i 's/\[2345\]/\[1\]/g' /etc/init/start-ttys.conf

# Enabling httpd
chkconfig --levels 2345 httpd on

# Disabling udev
mv /sbin/udevd /sbin/udevd_orig
chmod 444 /sbin/udevd_orig
ln -s /bin/true /sbin/udevd

echo "Last run on `date`" > /.CLEANRC


TIMImage_builder.sh
#!/bin/bash
#

# I want it to be verbose.
VERBOSE=false

##########################################################################
# Nothing to be changed below this point !
##########################################################################

# Programm Version
VER="$Revision: 1.0 $"

PATH=/sbin:/bin:/usr/sbin:/usr/bin:/usr/local/sbin:/usr/local/bin

# Get Program-Name, shortened Version.
PROGNAME="`basename $0 .sh`"

# Execution PID.
PROG_PID=$$

# Directory we work in.
BASEDIR=`pwd`

# Build Date in reverse - Prefix to builds
DATE=`date +"%Y%m%d"`
# Date + Time
LDATE=`date +"%F @ %T"`

# Lockfile
LockFile="${BASEDIR}/${PROGNAME}..LOCK"

# Define the Hostname
HOSTName=`hostname -s`
# Logfile - all info will go in there.
LogFile="${BASEDIR}/${DATE}_${PROGNAME}.log"
CFGFILE="${BASEDIR}/cfgs/${PROG_PID}.cfg"

# Old configuration provided as argument.
# Initiating non interactive mode
OLDCONF=$1

# Configuration file
SHAREMOD="${BASEDIR}/mod/share.inc"

if [ -f $SHAREMOD ]
then
    . $SHAREMOD
else
    echo "*** ERROR: Unable to load shared functions. Abort."
    exit
fi

##########################################################################
# Actual script start
##########################################################################

# Prevent double execution
Lock $LockFile 1

# Guest SPAN Interface
EXPOSEDPORTS="8080 8443 81 80"

# Create Title
title "Docker CA TIM Image Builder"
# Log program version.
log "$LDATE `whoami`@`hostname` - ${PROGNAME}.sh version $VER"

# In case we don't have a configuration file directory, we need to
# create it.
if [ ! -d cfgs ]
then
    MSG="Creating configuration file directory failed"
    mkdir cfgs
    errlvl=$?
    errors
fi

if [ -f "$OLDCONF" ]
then
    source $OLDCONF
    log "Using $OLDCONF presets"

else

    echo "# $LDATE `whoami`@`hostname` - ${PROGNAME}.sh version $VER" > $CFGFILE
    echo "# TIM Docker Image builder configuration file" >> $CFGFILE
    echo 'EXPOSEDPORTS="8080 8443 81 80"' >> $CFGFILE
    
    
    # This script will configure then build the TIM Docker image and
    # provide scripts to start and access the container
    echo ">>> Which TIM version do you want to integrate into the image ?"
    echo ">>> Available Versions are listed below:"
    echo "==============================================================="
    count=0
    for dir in `ls -d tim* 2> /dev/null`
    do
	if [ -d $dir ]
	then
	    echo " * $dir "
	    let count=($count + 1)
	fi
    done
    # Check if we actually have a TIM
    if [ $count -lt 1 ]
    then
	echo "No TIM Software directory was detected."
	echo "Please download the TIM Software from the official"
	echo "Download location (support.ca.com) and unpack the"
	echo "TIM Files into a Directory named after the TIM Version"
	echo "\"tim10.1\" for the CA APM TIM 10.1 release"
	MSG="No TIM detected. Abort!"
	errlvl=1
	errors
    fi
    
    echo "==============================================================="
    echo -n ">>> Please copy & paste the version to use: "
    read TIMVER
    while [ ! -x ${TIMVER}/timInstall.bin ]
    do
	echo -n "*** [Invalid entry !]: "
	read TIMVER
    done
    log "Chosen TIM Version: $TIMVER"
    echo "TIMVER=$TIMVER" >> $CFGFILE
    
    
    echo
    echo "==============================================================="
    echo -n ">>> How many TIM Worker processes should be activated [4]: "
    read WORKERtmp
    RESP=$WORKERtmp
    WORKER=${WORKERtmp:=4}
    MSG="Worker processes set to: $WORKER"
    log $MSG
    if [ -z "$RESP" ]
    then
	echo "*** $MSG"
    fi
    echo "WORKER=$WORKER" >> $CFGFILE
    
    echo
    echo "==============================================================="
    echo -n ">>> How do you want to name the Image [caapm/$TIMVER]: "
    read NAMEtmp
    RESP=$NAMEtmp
    IMAGENAME=${NAMEtmp:=caapm/$TIMVER}
    MSG="Image Name set to: $IMAGENAME"
    log $MSG
    if [ -z "$RESP" ]
    then
	echo "*** $MSG"
    fi
    echo "IMAGENAME=$IMAGENAME" >> $CFGFILE
    
    echo
    echo "==============================================================="
    echo "!!! You need to accept the Eula in ${TIMVER}/ca-eula.en.txt !"
    echo -n ">>> Do you accept the Eula [y/n]: "
    read EULA
    
    if [ "$EULA" != "y" ]
    then
	MSG="Eula not accepted. Aborting !"
	errlvl=1
	errors
    fi
    log "Eula Accepted ? : $EULA"
    echo "EULA=$EULA" >> $CFGFILE
    
    
    echo
    echo "==============================================================="
    echo -n ">>> How do you want to name the Container [$TIMVER]: "
    read CONTAINERNAMEtmp
    RESP=$CONTAINERNAMEtmp
    CONTAINERNAME=${CONTAINERNAMEtmp:=$TIMVER}
    MSG="Container Name set to: $CONTAINERNAME"
    log $MSG
    if [ -z "$RESP" ]
    then
	echo "*** $MSG"
    fi
    echo "CONTAINERNAME=$CONTAINERNAME" >> $CFGFILE

    echo
    echo "==============================================================="
    echo "Container Network mode"
    echo "Depending on the location the container will be installed, it can"
    echo "be necessary to separate the container from the running host."
    echo " - \"host\": all networking from the running host will be made"
    echo "    available to the inside of the container."
    echo " - \"secured\": the container will get its own network, and one"
    echo "    dedicated network interface will be hijacked from the host"
    echo "    and into the container as SPAN port."
    echo -n ">>> Container network mode [host|secured]: "
    read NMODEtmp
    RESP=$NMODEtmp
    NMODE=${NMODEtmp:=secured}
    MSG="Network mode set to: $NMODE"
    log $MSG
    if [ -z "$RESP" ]
    then
	echo "*** $MSG"
    fi
    echo "NMODE=$NMODE" >> $CFGFILE

    
    if [ "$NMODE" == "secured" ]
    then
	# Secured Mode - we can configure the docker proxy setup.

	echo
	echo "==============================================================="
	echo ">>> Which Docker Host physical network interface will be used "
	echo ">>> as SPAN Port. Note that the interface will be made exclusively"
	echo ">>> available to the container running the $TIMVER, and will not"
	echo ">>> be usable by the Docker Host or any other container !"
	echo ">>> Use ifconfig to identify a suitable interface !"
	echo -n "Which interface to use: [No Default]: "
	read HOSTDEV
	log "Host Device: $HOSTDEV"
	echo "HOSTDEV=$HOSTDEV" >> $CFGFILE
	
	
	echo 
	echo "==============================================================="
	echo "   For the following, keep in mind that the TIM uses by default"
	echo "   port 80, 81 8080 and 443. These are the ports the TIM will"
	echo "   run inside the Docker container. If any of these ports is "
	echo "   already in useon the host system, it cannot be used by the"
	echo "   docker-proxy."
	
	echo
	echo "==============================================================="
	echo -n ">>> Which port do you want to map the HTTP Port 80 [80]: "
	read PORTHTTPtmp
	RESP=$PORTHTTPtmp
	PORTHTTP=${PORTHTTPtmp:=80}
	MSG="Port 80 mapped to $PORTHTTP"
	log $MSG
	if [ -z "$RESP" ]
	then
	    echo "*** $MSG"
	fi
	echo "PORTHTTP=$PORTHTTP" >> $CFGFILE
	
	
	echo
	echo "==============================================================="
	echo -n ">>> Which port do you want to map the HTTP Port 8080 [8080]: "
	read PORTHTTP8tmp
	RESP=$PORTHTTP8tmp
	PORTHTTP8=${PORTHTTP8tmp:=8080}
	MSG="Port 8080 mapped to $PORTHTTP8"
	log $MSG
	if [ -z "$RESP" ]
	then
	    echo "*** $MSG"
	fi
	echo "PORTHTTP8=$PORTHTTP8" >> $CFGFILE
	
	echo
	echo "==============================================================="
	echo -n ">>> Which port do you want to map the HTTP Port 81 [81]: "
	read PORTHTTP1tmp
	RESP=$PORTHTTP1tmp
	PORTHTTP1=${PORTHTTP1tmp:=81}
	MSG="Port 81 mapped to $PORTHTTP1"
	log $MSG
	if [ -z "$RESP" ]
	then
	    echo "*** $MSG"
	fi
	echo "PORTHTTP1=$PORTHTTP1" >> $CFGFILE
	
	echo
	echo "==============================================================="
	echo -n ">>> Which port do you want to map the HTTPS Port 443 [8443]: "
	read PORTHTTPStmp
	RESP=$PORTHTTPStmp
	PORTHTTPS=${PORTHTTPStmp:=8443}
	MSG="Port 8443 mapped to $PORTHTTPS"
	log $MSG
	if [ -z "$RESP" ]
	then
	    echo "*** $MSG"
	fi
	echo "PORTHTTPS=$PORTHTTPS" >> $CFGFILE

    else

	# Host Mode - we can configure the TIM Ports.

	echo 
	echo "==============================================================="
	echo "   For the following, keep in mind that the TIM uses by default"
	echo "   port 80, 81 8080 and 443. You now have the possibility to"
	echo "   change these ports. Make sure you reflect these ports in"
	echo "   the rest of the setup (EM/Collector/TESS etc.)"
	
	echo
	echo "==============================================================="
	echo -n ">>> Which port do you want to map the HTTP Port 80 [80]: "
	read PORTHTTPtmp
	RESP=$PORTHTTPtmp
	PORTHTTP=${PORTHTTPtmp:=80}
	MSG="Port 80 mapped to $PORTHTTP"
	log $MSG
	if [ -z "$RESP" ]
	then
	    echo "*** $MSG"
	fi
	echo "PORTHTTP=$PORTHTTP" >> $CFGFILE
	
	echo
	echo "==============================================================="
	echo -n ">>> Which port do you want to map the HTTP Port 81 [81]: "
	read PORTHTTP1tmp
	RESP=$PORTHTTP1tmp
	PORTHTTP1=${PORTHTTP1tmp:=81}
	MSG="Port 81 mapped to $PORTHTTP1"
	log $MSG
	if [ -z "$RESP" ]
	then
	    echo "*** $MSG"
	fi
	echo "PORTHTTP1=$PORTHTTP1" >> $CFGFILE
	
	echo
	echo "==============================================================="
	echo -n ">>> Which port do you want to map the HTTPS Port 443 [443]: "
	read PORTHTTPStmp
	RESP=$PORTHTTPStmp
	PORTHTTPS=${PORTHTTPStmp:=443}
	MSG="Port 443 mapped to $PORTHTTPS"
	log $MSG
	if [ -z "$RESP" ]
	then
	    echo "*** $MSG"
	fi
	echo "PORTHTTPS=$PORTHTTPS" >> $CFGFILE

    fi
	
    MSG="Moving Configuration file to new name failed"
    mv $CFGFILE cfgs/${CONTAINERNAME}.cfg
    errlvl=$?
    errors

fi


# Reconfigure the exposed Ports
EXPOSEDPORTS="$PORTHTTP $PORTHTTP1 $PORTHTTPS $PORTHTTP8"


echo 
echo ">>> Options for this Docker Image Build"
echo "==============================================================="
# Required VARS in Dockerfile
echo -e "TIM Version:\t\t$TIMVER"
echo -e "TIM Workers:\t\t$WORKER"
echo -e "Image name:\t\t$IMAGENAME"
echo -e "Exposed ports:\t\t$EXPOSEDPORTS"
echo -e "Network mode:\t\t$NMODE"
# Required VARS in start_tim.sh
echo -e "Container Name:\t\t$CONTAINERNAME"

if [ "$NMODE" == "secured" ]
then
    echo -e "Host SPAN Int.:\t\t$HOSTDEV"
    echo -e "Port 80 mapped to:\t$PORTHTTP"
    echo -e "Port 81 mapped to:\t$PORTHTTP1"
    echo -e "Port 8080 mapped to:\t$PORTHTTP8"
    echo -e "Port 8443 mapped to:\t$PORTHTTPS"
else
    echo -e "Port 80 changed to:\t$PORTHTTP"
    echo -e "Port 81 changed to:\t$PORTHTTP1"
    echo -e "Port 8443 changed to:\t$PORTHTTPS"
fi
    
echo
echo "==============================================================="
echo -n ">>> Proceed [y/n]: "
read PROCEED
if [ "$PROCEED" != "y" ]
then
    MSG="User aborted process !"
    errlvl=1
    errors
fi

if [ "$NMODE" == "secured" ]
then

    # ===================================
    echo -n "*** Generating CA_AUTOMATION file: "
    MSG="Creating CA_AUTOMATION file failed"
    echo "export PRIVATE_HTTPD=0" > CA_AUTOMATION
    errlvl=$?
    echo "export HTTP_PORT=80" >> CA_AUTOMATION
    echo "export HTTPS_PORT=8443" >> CA_AUTOMATION
    echo "export TIM_PORT=81" >> CA_AUTOMATION
    echo "export MOD_FIREWALL=0" >> CA_AUTOMATION
    echo "export INSTALL_TYPE=F" >> CA_AUTOMATION
    echo "export UPGRADE_KEEP_SETTINGS=0" >> CA_AUTOMATION
    echo " OK. "

    DOCKERFILE=Dockerfile.tpl.secure
    
else

    # ===================================
    echo -n "*** Generating CA_AUTOMATION file: "
    MSG="Creating CA_AUTOMATION file failed"
    echo "export PRIVATE_HTTPD=0" > CA_AUTOMATION
    errlvl=$?
    echo "export HTTP_PORT=$PORTHTTP" >> CA_AUTOMATION
    echo "export HTTPS_PORT=$PORTHTTPS" >> CA_AUTOMATION
    echo "export TIM_PORT=$PORTHTTP1" >> CA_AUTOMATION
    echo "export MOD_FIREWALL=0" >> CA_AUTOMATION
    echo "export INSTALL_TYPE=F" >> CA_AUTOMATION
    echo "export UPGRADE_KEEP_SETTINGS=0" >> CA_AUTOMATION
    echo " OK. "

    DOCKERFILE=Dockerfile.tpl.host
    
fi
  
# ===================================
echo -n "*** Generating Dockerfile: "
MSG="Creating Dockerfile failed"
sed -e "s/TIMVER/$TIMVER/g" mod/${DOCKERFILE} > Dockerfile.tmp
errlvl=$?
errors
sed -i "s/EXPOSEDPORTS/$EXPOSEDPORTS/g" Dockerfile.tmp
errlvl=$?
errors
sed -i "s/WORKER/$WORKER/g" Dockerfile.tmp
errlvl=$?
errors
sed -i "s/PORTHTTPS/$PORTHTTPS/g" Dockerfile.tmp
errlvl=$?
errors
sed -i "s/PORTHTTP/$PORTHTTP/g" Dockerfile.tmp
errlvl=$?
errors


# Special case for TIM 9.6.1 or 9.6.0
if [ `echo $TIMVER | grep -c "^tim9.6"` -gt 0 ]
then
    sed -i 's#RUN /root/timInstallRPM.sh#RUN /root/timInstallRPM.sh ; exit 0#g' Dockerfile.tmp
fi
mv Dockerfile.tmp Dockerfile
errlvl=$?
errors
echo " OK. "

# ===================================
if [ "$NMODE" == "secured" ]
then
    # Secured network mode
    echo -n "*** Creating Container start script: "
    MSG="Creating start_tim.sh script failed"
    sed -e s#CONTAINERNAME#$CONTAINERNAME#g mod/start_tim_secured.tpl > start_tim.tmp
    errlvl=$?
    errors
    sed -i s#IMAGENAME#$IMAGENAME#g start_tim.tmp
    errlvl=$?
    errors
    sed -i s#PORTHTTP1#$PORTHTTP1#g start_tim.tmp
    errlvl=$?
    errors
    sed -i s#PORTHTTP8#$PORTHTTP8#g start_tim.tmp
    errlvl=$?
    errors
    sed -i s#PORTHTTPS#$PORTHTTPS#g start_tim.tmp
    errlvl=$?
    errors
    sed -i s#PORTHTTP#$PORTHTTP#g start_tim.tmp
    errlvl=$?
    errors
    sed -i s#HOSTDEV#$HOSTDEV#g start_tim.tmp
    errlvl=$?
    errors
    mv start_tim.tmp start_${CONTAINERNAME}.sh
    errlvl=$?
    errors
    chmod 755 start_${CONTAINERNAME}.sh
    errlvl=$?
    errors
    echo " OK. "
else
    # Host network mode
    echo -n "*** Creating Container start script: "
    MSG="Creating start_tim.sh script failed"
    sed -e s#CONTAINERNAME#$CONTAINERNAME#g mod/start_tim_host.tpl > start_tim.tmp
    errlvl=$?
    errors
    sed -i s#IMAGENAME#$IMAGENAME#g start_tim.tmp
    errlvl=$?
    errors
    mv start_tim.tmp start_${CONTAINERNAME}.sh
    errlvl=$?
    errors
    chmod 755 start_${CONTAINERNAME}.sh
    errlvl=$?
    errors
    echo " OK. "

fi
# ===================================
echo -n "*** Creating Container access script: "
MSG="Creating tim_shell.sh script failed"
echo "docker exec -it $CONTAINERNAME /bin/bash" > ${CONTAINERNAME}_shell.sh
errlvl=$?
errors
chmod 755 ${CONTAINERNAME}_shell.sh
errlvl=$?
errors
echo " OK. "

echo
echo "==============================================================="
echo -n ">>> Create image [y/n]: "
read DOIMAGE
if [ "$DOIMAGE" != "y" ]
then
    MSG="User aborted process !"
    errlvl=1
    errors
fi
MSG="Creating Image $IMAGENAME"
log MSG
docker build -t="$IMAGENAME" .
errlvl=$?
errors


Unlock $LockFile
